### PR TITLE
[Serve] Latency improvement by using pickle

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
For performance reasons, this PR pickles `request_metadata` and `HTTPRequestWrapper`.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -304,8 +304,8 @@ cdef prepare_args(
         else:
             serialized_arg = worker.get_serialization_context().serialize(arg)
             metadata = serialized_arg.metadata
-            metadata_fields = metadata.split(b",")
             if language != Language.PYTHON:
+                metadata_fields = metadata.split(b",")
```
Comment (simon-mo) [inline python/ray/_raylet.pyx:308]: a tiny perf improvement on the critical path. `metadata_fields` is only used in this (almost never hit) branch

File: python/ray/serve/handle.py [hunk 2]
```diff
@@ -105,12 +107,15 @@ def _make_router(self) -> EndpointRouter:
...
-                http_method: Union[str, DEFAULT] = DEFAULT.VALUE,
-                http_headers: Union[Dict[str, str], DEFAULT] = DEFAULT.VALUE):
+    def options(
+            self,
+            *,
+            method_name: Union[str, DEFAULT] = DEFAULT.VALUE,
+            shard_key: Union[str, DEFAULT] = DEFAULT.VALUE,
+            http_method: Union[str, DEFAULT] = DEFAULT.VALUE,
+            http_headers: Union[Dict[str, str], DEFAULT] = DEFAULT.VALUE,
+            http_arg_is_pickled: Union[bool, DEFAULT] = DEFAULT.VALUE,
```
Comment (edoakes) [inline python/ray/serve/handle.py:117]: this should not be user-facing

File: python/ray/serve/utils.py [hunk 3]
```diff
@@ -93,6 +94,10 @@ def parse_request_item(request_item):
             return (arg, ), {}
         elif isinstance(arg, HTTPRequestWrapper):
             return (build_starlette_request(arg.scope, arg.body), ), {}
```
Comment (edoakes) [inline python/ray/serve/utils.py:96]: we don't need this condition anymore, right?

**Comment by simon-mo** [comment]
@edoakes updated.

**Review by edoakes — APPROVED** [review]

**Pull request merged**